shell/platform/android/surface/snapshot_surface_producer.h

```diff
@@ -0,0 +1,29 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef SHELL_PLATFORM_ANDROID_SURFACE_SNAPSHOT_SURFACE_PRODUCER_H_
+#define SHELL_PLATFORM_ANDROID_SURFACE_SNAPSHOT_SURFACE_PRODUCER_H_
+
+#include <memory>
+
+#include "flutter/flow/surface.h"
+#include "flutter/shell/common/snapshot_surface_producer.h"
+#include "flutter/shell/platform/android/surface/android_surface.h"
+
+namespace flutter {
+
+class AndroidSnapshotSurfaceProducer : public SnapshotSurfaceProducer {
+ public:
+  explicit AndroidSnapshotSurfaceProducer(
+      std::shared_ptr<AndroidSurface> android_surface);
+
+  // |SnapshotSurfaceProducer|
+  std::unique_ptr<Surface> CreateSnapshotSurface() override;
+
+ private:
+  std::shared_ptr<AndroidSurface> android_surface_;
```

Comment:
I think this should hold a reference to the `AndroidSurface` instead of converting `AndroidSurface` into a shared pointer.
(similar to the `AndroidContext&` held by the `AndroidExternalViewEmbedder`)

I don't think it's safe to delete the last `AndroidSurface` pointer on a thread other than the platform thread.

Normally this would not happen because the `Shell` would delete the `Rasterizer` before the `PlatformView`, which would ensure that the `PlatformView`'s pointer to the `AndroidSurface` would be the last to go.  But if the `PlatformView` must be responsible for deleting the `AndroidSurface`, then I'd prefer to formalize that by having the `PlatformView` hold a `unique_ptr`.

Alternatively, `CreateSnapshotSurface` could be moved out of `AndroidSurface` and into `AndroidContext`.  The `AndroidSurfaceGL::CreatePbufferSurface` implementation is delegating to `AndroidContextGL` and does not use any of `AndroidSurfaceGL`'s state.

Reply:
We do need an AndroidSurfaceGL to create a GPUSurfaceGL which is what the rasterizer expects to interact with. I'll ook at using the ref.

Reply:
Using a ref is less than ideal because it access a bunch of non-const methods. 

I'm using a weak_ptr which I think should be safe w.r.t. your concerns.

Reply:
Ok - weak_ptr doesn't actually solve this problem, because we'd still potentially release the shared_ptr on the wrong thread. changed to a non-const reference.